File: app/src/main/java/com/example/boardexamreviewer/data/AppDao.java
package com.example.boardexamreviewer.data;

import androidx.room.*;
import java.util.List;

@Dao
public interface AppDao {
    // --- Documents ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertDocument(DocumentEntity document);

    @Query("SELECT * FROM documents WHERE userId = :userId ORDER BY timestamp DESC")
    List<DocumentEntity> getAllDocumentsByUser(int userId);

    @Query("SELECT * FROM documents WHERE userId = :userId ORDER BY timestamp DESC LIMIT 1")
    DocumentEntity getLastDocumentByUser(int userId);

    @Delete
    void deleteDocument(DocumentEntity document);

    // --- Reviewers ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertReviewer(ReviewerEntity reviewer);

    @Query("SELECT * FROM reviewers WHERE userId = :userId ORDER BY timestamp DESC")
    List<ReviewerEntity> getAllReviewersByUser(int userId);

    // --- Quizzes ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertQuiz(QuizEntity quiz);

    @Query("SELECT * FROM quizzes WHERE userId = :userId ORDER BY timestamp DESC")
    List<QuizEntity> getAllQuizzesByUser(int userId);

    // --- Users (Profile Module) ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertUser(UserEntity user);

    @Query("SELECT * FROM users ORDER BY name ASC")
    List<UserEntity> getAllUsers();

    @Query("SELECT * FROM users WHERE id = :userId")
    UserEntity getUserById(int userId);

    // --- Categories ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertCategory(CategoryEntity category);

    @Query("SELECT * FROM categories ORDER BY name ASC")
    List<CategoryEntity> getAllCategories();

    // --- Progress Tracking ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertStudySession(StudySessionEntity session);

    @Query("SELECT * FROM study_sessions WHERE userId = :userId ORDER BY timestamp DESC")
    List<StudySessionEntity> getStudySessionsByUser(int userId);

    @Query("SELECT SUM(durationMinutes) FROM study_sessions WHERE userId = :userId AND sessionType = 'Work'")
    Integer getTotalStudyTime(int userId);
}

File: app/src/main/java/com/example/boardexamreviewer/data/AppDao.kt
package com.example.boardexamreviewer.data

import androidx.room.*

@Dao
interface AppDao {
    // --- Documents ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertDocument(document: DocumentEntity): Long

    @Query("SELECT * FROM documents WHERE userId = :userId ORDER BY timestamp DESC")
    suspend fun getAllDocumentsByUser(userId: Int): List<DocumentEntity>

    @Query("SELECT * FROM documents WHERE userId = :userId ORDER BY timestamp DESC LIMIT 1")
    suspend fun getLastDocumentByUser(userId: Int): DocumentEntity?

    @Delete
    suspend fun deleteDocument(document: DocumentEntity)

    // --- Reviewers ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertReviewer(reviewer: ReviewerEntity)

    @Query("SELECT * FROM reviewers WHERE userId = :userId ORDER BY timestamp DESC")
    suspend fun getAllReviewersByUser(userId: Int): List<ReviewerEntity>

    // --- Quizzes ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertQuiz(quiz: QuizEntity)

    @Query("SELECT * FROM quizzes WHERE userId = :userId ORDER BY timestamp DESC")
    suspend fun getAllQuizzesByUser(userId: Int): List<QuizEntity>

    // --- Users (Profile Module) ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertUser(user: UserEntity): Long

    @Query("SELECT * FROM users ORDER BY name ASC")
    suspend fun getAllUsers(): List<UserEntity>

    @Query("SELECT * FROM users WHERE id = :userId")
    suspend fun getUserById(userId: Int): UserEntity?

    // --- Categories ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertCategory(category: CategoryEntity): Long

    @Query("SELECT * FROM categories ORDER BY name ASC")
    suspend fun getAllCategories(): List<CategoryEntity>

    // --- Progress Tracking ---
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertStudySession(session: StudySessionEntity)

    @Query("SELECT * FROM study_sessions WHERE userId = :userId ORDER BY timestamp DESC")
    suspend fun getStudySessionsByUser(userId: Int): List<StudySessionEntity>

    @Query("SELECT SUM(durationMinutes) FROM study_sessions WHERE userId = :userId AND sessionType = 'Work'")
    suspend fun getTotalStudyTime(userId: Int): Int?
}
